### PR TITLE
I85

### LINKED ISSUE #85: Make use of automated testing techniques.
utilities:
- golang fuzzing https://github.com/google/gofuzz
- golang https://golang.org/pkg/net/http/httptest/
- golang https://golang.org/pkg/testing/

goals:
- fuzz data
- unit test

This also closes #69

### PR DESCRIPTION
**Pull request by gevorgyana**
Working yet.

closes #85

### CODE REVIEW

File: web-service/main.go [hunk 1]
```diff
@@ -1,50 +1,14 @@
...
-
-	db := containers.NewDB()
-	if err := db.Initialize(config.Internal.DbPath); err != nil {
-		errorLogger.Fatalln("Unable to open db at", config.Internal.DbPath, err)
-	}
-
-	api.InitializeControllers(db)
-	s3support.InitializeS3Support()
-}
-
```
Comment (gevorgyana) [inline web-service/main.go:47]: We can't have this thing in the main package, as we should import the initialization logic in unit tests (means linear complexity - for each unit!), and we can't import main package.

File: web-service/main.go [hunk 2]
```diff
@@ -53,10 +17,10 @@ func setupRoutes() {
...
 func main() {
-	configure()
+	initialize.Configure()
 
 	setupRoutes()
 
-	debugLogger.Println("Starting fair online judge service on", config.Server.Port)
-	errorLogger.Fatalln(http.ListenAndServe(config.Server.Port, nil))
+	utils.DebugLogger.Println("Starting fair online judge service on", config.Server.Port)
+	utils.ErrorLogger.Fatalln(http.ListenAndServe(config.Server.Port, nil))
```
Comment (gevorgyana) [inline web-service/main.go:25]: It was really excessive to have vars declared in this module. They were referring to `utils`-loggers anyway.

**Review by gevorgyana — COMMENTED** [review]
A small comment on unit tests. Just read this

А как юнит тесты лучше организовать? В Расте пишту в подмодуле каждого модуля. В гоу тоже так принято? Ну тут подмодулей не будет, но я имею в виду как-то так организовать тесты

package.go
test/package_test.go

И так для каждого модуля.

---

package.go и package_test.go спокойно живут в одной папке, но у _test имя пакета может отличаться от основного

**Comment by gevorgyana** [comment]
closes #69 and #85

File: web-service/src/server/server.go [hunk 3]
```diff
@@ -66,3 +66,23 @@ func NewServer(
 
 	return s
 }
+
+func NewTestServer(
+	conf *config.Configuration) *Server {
```
Comment (Pilipets) [inline web-service/src/server/server.go:71]: Why is this in the production code?
It shouldn't be there: we can ignore the building of testing files for the release build, but this method interferes with that idea.

File: web-service/src/config/config.go [hunk 4]
```diff
@@ -7,6 +7,10 @@ import (
 	utils "web-service/src/utils"
 )
 
+const (
+	ConfigPath = "config/main_config.json"
+)
```
Comment (Pilipets) [inline web-service/src/config/config.go:12]: What for?

File: web-service/src/s3support/fuzz_test.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+	minio, err = s3support.NewMinioService(&conf.Minio)
+	if err != nil {
+		t.Fatal("Unable to initialize s3Client", err)
+	}
+
+	t.Run("Fuzz", TestFuzz)
+}
+
+func TestFuzz(t *testing.T) {
```
Comment (Pilipets) [inline web-service/src/s3support/fuzz_test.go:45]: Please, write a meaningful name for the test, so one won't strive to guess what it does.

File: web-service/src/server/api_test.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+	config "web-service/src/config"
+)
+
+var server *http_server.Server
+
+const (
+	configPath = "../../config/main_config.json"
+)
+
+func UploadFilesActivityWorksInNormalUseCase(t *testing.T) {
```
Comment (Pilipets) [inline web-service/src/server/api_test.go:22]: This looks like the test with one file only, so designate it somehow in the method name.

**Review by Pilipets — CHANGES_REQUESTED** [review]

File: web-service/src/server/server.go (hunk 3 above)
Comment (gevorgyana) [reply web-service/src/server/server.go:71]: I want a suggestion then. How do we test the handlers in isolation? We can't access them without creating an instance of this type.

File: web-service/src/server/api_test.go (hunk 6 above)
Comment (gevorgyana) [reply web-service/src/server/api_test.go:22]: The name is `$TESTED_FUNCTION_NAME + SomeDescriptiveLogicForDeveloper`.

**Review by Pilipets — DISMISSED** [review]
Good start for testing.

**Comment by Pilipets** [comment]
Please, don't forget to squash two commits into one when merging.

File: web-service/src/server/server.go (hunk 3 above)
Comment (Pilipets) [reply web-service/src/server/server.go:71]: We don't need this method at all - it just passes nil dependencies.
You can use `func NewServer(...)` for that purpose.

**Comment by Pilipets** [comment]
Remove `func NewTestServer(...)` and use `func NewServer(...)` with the respective parameters instead.

**Review by Pilipets — APPROVED** [review]

**Pull request merged**